# HG changeset patch
# User Carlo de Falco <cdf@users.sourceforge.net>
# Date 1341141050 -7200
#      Sun Jul 01 13:10:50 2012 +0200
# Node ID 93e5ade3fda43a14aef811a3224a4086cd41a162
# Parent  aad7d7ee8e8adf16939532227d94e6ad12e7cfa6
Improve matlab compatibility of gmres

* scripts/sparse/gmres.m: adapt resvec, iter and relres outputs
to be more compatible with matlab

diff --git a/scripts/sparse/gmres.m b/scripts/sparse/gmres.m
--- a/scripts/sparse/gmres.m
+++ b/scripts/sparse/gmres.m
@@ -169,35 +169,34 @@ function [x, flag, relres, it, resvec] =
 
     little_res = B(1:restart_it+1) - ...
         H(1:restart_it+1, 1:restart_it) * Y(1:restart_it);
 
     presn = norm (little_res, 2);
 
     x = x_old + V(:, 1:restart_it) * Y(1:restart_it);
 
-    resvec(iter) = presn;
+    resvec(iter+1) = presn;
     if (norm (x - x_old, inf) <= eps)
       flag = 3;  # Stagnation: no change between iterations
       break;
     endif
 
     restart_it++ ;
     iter++;
   endwhile
 
   if (nargout > 1)
     ## Calculate extra outputs as requested
     relres = presn / prec_b_norm;
     if (relres <= rtol)
       flag = 0;  # Converged to solution within tolerance
     endif
 
-    resvec = resvec(1:iter-1);
-    it = [ceil(iter / restart), rem(iter, restart)];
+    it = [floor(iter/restart), restart_it-1];
   endif
 
 endfunction
 
 
 %!demo
 %! dim = 20;
 %! A = spdiags ([-ones(dim,1) 2*ones(dim,1) ones(dim,1)], [-1:1], dim, dim);
